<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.10.0" />
<title>src.package.bamoperations.bamoperations API documentation</title>
<meta name="description" content="A collection of classes or functions that performs bam processing operations" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>src.package.bamoperations.bamoperations</code></h1>
</header>
<section id="section-intro">
<p>A collection of classes or functions that performs bam processing operations</p>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">&#34;&#34;&#34;
A collection of classes or functions that performs bam processing operations
&#34;&#34;&#34;

import os.path
import pysam

from package.datastructureoperations.listoperations.listhandlers import (get_first_element, get_second_element,
                                                                         get_third_element)
from package.enums.delimiter_enums import Delimiters
from package.fileoperations.filehandlers import read_csv
from package.fileoperations.filewriters import FileWriter


class BamOperator:
    &#34;&#34;&#34;Performs bam processing operations&#34;&#34;&#34;

    bam_files_directory = None
    bed_file = None
    output_directory = None

    def __init__(self, bam_files: list, bed_file: str, output_directory: str) -&gt; None:
        &#34;&#34;&#34;
        Constructor

        Parameters
        ----------
        bam_files :list
           Path to bam files
        bed_file:str
           Path to bwa meth bam file
        output_directory:str
           Path to output directory
        &#34;&#34;&#34;

        self.bam_files = bam_files
        self.bed_file = bed_file
        self.output_directory = output_directory

    def process_bam_files(self) -&gt; None:
        &#34;&#34;&#34;process bam files &#34;&#34;&#34;

        bed_file_content: list[str] = read_csv(self.bed_file, delimiter=Delimiters.TAB_SEPERATOR)

        for count, bam_file in enumerate(self.bam_files):

            output_read_counts_sequences: tuple = BamOperator.__process_bam_file(bam_file, bed_file_content)

            output_read_counts: dict = get_first_element(output_read_counts_sequences)

            fasta_sequences: str = BamOperator.__generate_fasta(get_second_element(output_read_counts_sequences))

            BamOperator.__write_to_out(output_read_counts, fasta_sequences,
                                       BamOperator.__prepare_json_filenames(count, self.output_directory),
                                       BamOperator.__prepare_fasta_filenames(count, self.output_directory))

    @classmethod
    def __prepare_json_filenames(cls, count: int, output_directory) -&gt; str:
        &#34;&#34;&#34; prepare json filenames &#34;&#34;&#34;

        __file_basename_json: str = str(count).join([&#34;read_counts&#34;, &#34;.json&#34;])

        json_output_file: str = os.path.join(output_directory, __file_basename_json)

        return json_output_file

    @classmethod
    def __prepare_fasta_filenames(cls, count: int, output_directory) -&gt; str:
        &#34;&#34;&#34; prepare fasta filenames &#34;&#34;&#34;

        __file_basename_fasta: str = str(count).join([&#34;fasta_sequence&#34;, &#34;.fasta&#34;])

        fasta_output_file: str = os.path.join(output_directory, __file_basename_fasta)

        return fasta_output_file

    @classmethod
    def __write_to_out(cls, output_read_counts: dict, fasta_sequences: str,
                       json_output_file: str, fasta_output_file: str) -&gt; None:
        &#34;&#34;&#34;Writes the output to the output file. &#34;&#34;&#34;

        file_writer_json: FileWriter = FileWriter(json_output_file, &#34;w&#34;)

        file_writer_json.write_json(output_read_counts)

        file_writer_fasta: FileWriter = FileWriter(fasta_output_file, &#34;w&#34;)

        file_writer_fasta.write_str(fasta_sequences)

    @classmethod
    def __generate_fasta(cls, read_ids_sequences: list) -&gt; str:
        &#34;&#34;&#34; Generate fasta file from read_ids_sequences &#34;&#34;&#34;

        fasta_sequences: list = [
            Delimiters.FASTA_IDENTIFIER + Delimiters.NEW_LINER.join([read_query_id, read_sequence])
            for element in read_ids_sequences
            for read_query_id, read_sequence in
            element.items()
        ]

        return Delimiters.NEW_LINER.join(fasta_sequences)

    @classmethod
    def __process_bam_file(cls, bam_file: str, bed_file_content: list[str]) -&gt; tuple:
        &#34;&#34;&#34; process each bam file &#34;&#34;&#34;

        with pysam.AlignmentFile(bam_file, &#34;rb&#34;) as bam_file_content:
            return BamOperator.__process_bed_file_content(bed_file_content, bam_file_content)

    @classmethod
    def __process_bed_file_content(cls, bed_file_content: list, bam_file_content: object()) -&gt; tuple:
        &#34;&#34;&#34; process bed file content &#34;&#34;&#34;

        read_interval_counts = {}
        read_sequences = []

        for line in bed_file_content:
            chromosome, start, end = get_first_element(line), get_second_element(line), get_third_element(line)
            start, end = int(start), int(end)

            read_names_sequence_pairs: dict = BamOperator.__process_bed_line(line, bam_file_content)

            read_counts_key: str = Delimiters.HYPHEN.join([chromosome, str(start), str(end)])

            read_interval_counts[read_counts_key] = len(list(read_names_sequence_pairs.keys()))

            read_sequences.append(read_names_sequence_pairs)

        return read_interval_counts, read_sequences

    @classmethod
    def __process_bed_line(cls, line: str, bam_file_content: object()) -&gt; dict:
        &#34;&#34;&#34; process bed file &#34;&#34;&#34;

        chromosome, start, end = get_first_element(line), get_second_element(line), get_third_element(line)
        start, end = int(start), int(end)

        reads: dict = {}

        for read in bam_file_content:

            if read.is_mapped and read.reference_name == chromosome:
                if BamOperator.__is_interval(read, start, end):
                    reads[read.query_name] = read.query_sequence.upper()

        return reads

    @classmethod
    def __is_interval(cls, read: object(), bed_start: int, bed_end: int) -&gt; bool:
        &#34;&#34;&#34; check if read is within interval &#34;&#34;&#34;

        return read.reference_start &gt;= bed_start and read.reference_end &lt;= bed_end</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="src.package.bamoperations.bamoperations.BamOperator"><code class="flex name class">
<span>class <span class="ident">BamOperator</span></span>
<span>(</span><span>bam_files: list, bed_file: str, output_directory: str)</span>
</code></dt>
<dd>
<div class="desc"><p>Performs bam processing operations</p>
<p>Constructor</p>
<h2 id="parameters">Parameters</h2>
<p>bam_files :list
Path to bam files
bed_file:str
Path to bwa meth bam file
output_directory:str
Path to output directory</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class BamOperator:
    &#34;&#34;&#34;Performs bam processing operations&#34;&#34;&#34;

    bam_files_directory = None
    bed_file = None
    output_directory = None

    def __init__(self, bam_files: list, bed_file: str, output_directory: str) -&gt; None:
        &#34;&#34;&#34;
        Constructor

        Parameters
        ----------
        bam_files :list
           Path to bam files
        bed_file:str
           Path to bwa meth bam file
        output_directory:str
           Path to output directory
        &#34;&#34;&#34;

        self.bam_files = bam_files
        self.bed_file = bed_file
        self.output_directory = output_directory

    def process_bam_files(self) -&gt; None:
        &#34;&#34;&#34;process bam files &#34;&#34;&#34;

        bed_file_content: list[str] = read_csv(self.bed_file, delimiter=Delimiters.TAB_SEPERATOR)

        for count, bam_file in enumerate(self.bam_files):

            output_read_counts_sequences: tuple = BamOperator.__process_bam_file(bam_file, bed_file_content)

            output_read_counts: dict = get_first_element(output_read_counts_sequences)

            fasta_sequences: str = BamOperator.__generate_fasta(get_second_element(output_read_counts_sequences))

            BamOperator.__write_to_out(output_read_counts, fasta_sequences,
                                       BamOperator.__prepare_json_filenames(count, self.output_directory),
                                       BamOperator.__prepare_fasta_filenames(count, self.output_directory))

    @classmethod
    def __prepare_json_filenames(cls, count: int, output_directory) -&gt; str:
        &#34;&#34;&#34; prepare json filenames &#34;&#34;&#34;

        __file_basename_json: str = str(count).join([&#34;read_counts&#34;, &#34;.json&#34;])

        json_output_file: str = os.path.join(output_directory, __file_basename_json)

        return json_output_file

    @classmethod
    def __prepare_fasta_filenames(cls, count: int, output_directory) -&gt; str:
        &#34;&#34;&#34; prepare fasta filenames &#34;&#34;&#34;

        __file_basename_fasta: str = str(count).join([&#34;fasta_sequence&#34;, &#34;.fasta&#34;])

        fasta_output_file: str = os.path.join(output_directory, __file_basename_fasta)

        return fasta_output_file

    @classmethod
    def __write_to_out(cls, output_read_counts: dict, fasta_sequences: str,
                       json_output_file: str, fasta_output_file: str) -&gt; None:
        &#34;&#34;&#34;Writes the output to the output file. &#34;&#34;&#34;

        file_writer_json: FileWriter = FileWriter(json_output_file, &#34;w&#34;)

        file_writer_json.write_json(output_read_counts)

        file_writer_fasta: FileWriter = FileWriter(fasta_output_file, &#34;w&#34;)

        file_writer_fasta.write_str(fasta_sequences)

    @classmethod
    def __generate_fasta(cls, read_ids_sequences: list) -&gt; str:
        &#34;&#34;&#34; Generate fasta file from read_ids_sequences &#34;&#34;&#34;

        fasta_sequences: list = [
            Delimiters.FASTA_IDENTIFIER + Delimiters.NEW_LINER.join([read_query_id, read_sequence])
            for element in read_ids_sequences
            for read_query_id, read_sequence in
            element.items()
        ]

        return Delimiters.NEW_LINER.join(fasta_sequences)

    @classmethod
    def __process_bam_file(cls, bam_file: str, bed_file_content: list[str]) -&gt; tuple:
        &#34;&#34;&#34; process each bam file &#34;&#34;&#34;

        with pysam.AlignmentFile(bam_file, &#34;rb&#34;) as bam_file_content:
            return BamOperator.__process_bed_file_content(bed_file_content, bam_file_content)

    @classmethod
    def __process_bed_file_content(cls, bed_file_content: list, bam_file_content: object()) -&gt; tuple:
        &#34;&#34;&#34; process bed file content &#34;&#34;&#34;

        read_interval_counts = {}
        read_sequences = []

        for line in bed_file_content:
            chromosome, start, end = get_first_element(line), get_second_element(line), get_third_element(line)
            start, end = int(start), int(end)

            read_names_sequence_pairs: dict = BamOperator.__process_bed_line(line, bam_file_content)

            read_counts_key: str = Delimiters.HYPHEN.join([chromosome, str(start), str(end)])

            read_interval_counts[read_counts_key] = len(list(read_names_sequence_pairs.keys()))

            read_sequences.append(read_names_sequence_pairs)

        return read_interval_counts, read_sequences

    @classmethod
    def __process_bed_line(cls, line: str, bam_file_content: object()) -&gt; dict:
        &#34;&#34;&#34; process bed file &#34;&#34;&#34;

        chromosome, start, end = get_first_element(line), get_second_element(line), get_third_element(line)
        start, end = int(start), int(end)

        reads: dict = {}

        for read in bam_file_content:

            if read.is_mapped and read.reference_name == chromosome:
                if BamOperator.__is_interval(read, start, end):
                    reads[read.query_name] = read.query_sequence.upper()

        return reads

    @classmethod
    def __is_interval(cls, read: object(), bed_start: int, bed_end: int) -&gt; bool:
        &#34;&#34;&#34; check if read is within interval &#34;&#34;&#34;

        return read.reference_start &gt;= bed_start and read.reference_end &lt;= bed_end</code></pre>
</details>
<h3>Class variables</h3>
<dl>
<dt id="src.package.bamoperations.bamoperations.BamOperator.bam_files_directory"><code class="name">var <span class="ident">bam_files_directory</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="src.package.bamoperations.bamoperations.BamOperator.bed_file"><code class="name">var <span class="ident">bed_file</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="src.package.bamoperations.bamoperations.BamOperator.output_directory"><code class="name">var <span class="ident">output_directory</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
</dl>
<h3>Methods</h3>
<dl>
<dt id="src.package.bamoperations.bamoperations.BamOperator.process_bam_files"><code class="name flex">
<span>def <span class="ident">process_bam_files</span></span>(<span>self) ‑> None</span>
</code></dt>
<dd>
<div class="desc"><p>process bam files</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def process_bam_files(self) -&gt; None:
    &#34;&#34;&#34;process bam files &#34;&#34;&#34;

    bed_file_content: list[str] = read_csv(self.bed_file, delimiter=Delimiters.TAB_SEPERATOR)

    for count, bam_file in enumerate(self.bam_files):

        output_read_counts_sequences: tuple = BamOperator.__process_bam_file(bam_file, bed_file_content)

        output_read_counts: dict = get_first_element(output_read_counts_sequences)

        fasta_sequences: str = BamOperator.__generate_fasta(get_second_element(output_read_counts_sequences))

        BamOperator.__write_to_out(output_read_counts, fasta_sequences,
                                   BamOperator.__prepare_json_filenames(count, self.output_directory),
                                   BamOperator.__prepare_fasta_filenames(count, self.output_directory))</code></pre>
</details>
</dd>
</dl>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="src.package.bamoperations" href="index.html">src.package.bamoperations</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="src.package.bamoperations.bamoperations.BamOperator" href="#src.package.bamoperations.bamoperations.BamOperator">BamOperator</a></code></h4>
<ul class="">
<li><code><a title="src.package.bamoperations.bamoperations.BamOperator.bam_files_directory" href="#src.package.bamoperations.bamoperations.BamOperator.bam_files_directory">bam_files_directory</a></code></li>
<li><code><a title="src.package.bamoperations.bamoperations.BamOperator.bed_file" href="#src.package.bamoperations.bamoperations.BamOperator.bed_file">bed_file</a></code></li>
<li><code><a title="src.package.bamoperations.bamoperations.BamOperator.output_directory" href="#src.package.bamoperations.bamoperations.BamOperator.output_directory">output_directory</a></code></li>
<li><code><a title="src.package.bamoperations.bamoperations.BamOperator.process_bam_files" href="#src.package.bamoperations.bamoperations.BamOperator.process_bam_files">process_bam_files</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.10.0</a>.</p>
</footer>
</body>
</html>